File: docs-examples/example-java/src/test/java/io/micronaut/rabbitmq/docs/consumer/custom/annotation/DeliveryTagAnnotationBinder.java
package io.micronaut.rabbitmq.docs.consumer.custom.annotation;

import io.micronaut.context.annotation.Requires;
// tag::imports[]
import io.micronaut.core.convert.ArgumentConversionContext;
import io.micronaut.core.convert.ConversionService;
import io.micronaut.rabbitmq.bind.RabbitAnnotatedArgumentBinder;
import io.micronaut.rabbitmq.bind.RabbitConsumerState;
import jakarta.inject.Singleton;
// end::imports[]

@Requires(property = "spec.name", value = "DeliveryTagSpec")
// tag::clazz[]
@Singleton // <1>
public class DeliveryTagAnnotationBinder implements RabbitAnnotatedArgumentBinder<DeliveryTag> { // <2>

    private final ConversionService<?> conversionService;

    public DeliveryTagAnnotationBinder(ConversionService<?> conversionService) { // <3>
        this.conversionService = conversionService;
    }

    @Override
    public Class<DeliveryTag> getAnnotationType() {
        return DeliveryTag.class;
    }

    @Override
    public BindingResult<Object> bind(ArgumentConversionContext<Object> context, RabbitConsumerState source) {
        Long deliveryTag = source.getEnvelope().getDeliveryTag(); // <4>
        return () -> conversionService.convert(deliveryTag, context); // <5>
    }
}
// end::clazz[]

File: docs-examples/example-python/src/test/python/micronaut/rabbitmq/docs/consumer/custom/annotation/DeliveryTagAnnotationBinder.py
import java
from micronaut.context.annotation import Requires
# tag::imports[]
from jakarta.inject import Singleton
from micronaut.core.bind.ArgumentBinder import BindingResult
from micronaut.core.convert import ArgumentConversionContext, ConversionService
from micronaut.rabbitmq.bind import RabbitAnnotatedArgumentBinder, RabbitConsumerState

from .DeliveryTag import DeliveryTag
# end::imports[]

# TODO(python): java.type needed because the annotation type is returned to Java as a runtime java.lang.Class
# (RabbitAnnotatedArgumentBinder.getAnnotationType()); a Python-defined annotation function is not accepted
# ("Cannot convert '<function DeliveryTag>' (language: Python, type: function) to Java type 'java.lang.Class'")
DeliveryTagClass = java.type("micronaut.rabbitmq.docs.consumer.custom.annotation.DeliveryTag")


@Requires(property="spec.name", value="DeliveryTagSpec")
# tag::clazz[]
@Singleton  # <1>
class DeliveryTagAnnotationBinder(RabbitAnnotatedArgumentBinder[DeliveryTag]):  # <2>

    def __init__(self, conversion_service: ConversionService):  # <3>
        self.conversion_service = conversion_service

    def getAnnotationType(self) -> type[DeliveryTag]:
        return DeliveryTagClass

    def bind(self, context: ArgumentConversionContext[object], source: RabbitConsumerState) -> BindingResult[object]:
        delivery_tag = source.getEnvelope().getDeliveryTag()  # <4>
        return lambda: self.conversion_service.convert(delivery_tag, context)  # <5>
# end::clazz[]
